alg_convex_hull.go

```diff
@@ -1,110 +1,191 @@
 package simplefeatures
 
 import (
+	"fmt"
 	"sort"
 )
 
+type threePointOrientation int
+
 const (
 	// rightTurn indicates the orientation is right turn which is anticlockwise
-	rightTurn = iota
+	rightTurn threePointOrientation = iota + 1
 	// collinear indicates three points are on the same line
 	collinear
 	// leftTurn indicates the orientation is left turn which is clockwise
 	leftTurn
 )
 
-// graphamScan returns a convex hull.
+func (o threePointOrientation) String() string {
+	switch o {
+	case rightTurn:
+		return "right turn"
+	case collinear:
+		return "collinear"
+	case leftTurn:
+		return "left turn"
+	default:
+		return "invalid orientation"
+	}
+}
+
+// orientation checks if s is on the right hand side or left hand side of the line formed by p and q.
+func orientation(p, q, s XY) threePointOrientation {
+	cp := q.Sub(p).Cross(s.Sub(q))
+	switch {
+	case cp.GT(zero):
+		return leftTurn
+	case cp.LT(zero):
+		return rightTurn
+	default:
+		return collinear
+	}
+}
+
+func convexHull(g Geometry) Geometry {
+	if g.IsEmpty() {
+		// special case to mirror postgis behaviour
+		return g
+	}
+	pts := g.convexHullPointSet()
+	hull := grahamScan(pts)
+	switch len(hull) {
+	case 0:
+		return NewGeometryCollection(nil)
+	case 1:
+		return NewPoint(hull[0])
+	case 2:
+		ln, err := NewLine(
+			Coordinates{hull[0]},
+			Coordinates{hull[1]},
+		)
+		if err != nil {
+			panic("bug in grahamScan routine - output 2 coincident points")
+		}
+		return ln
```

Comment:
Not quite sure why do you need to take into account the situation that `len(hull)` is less than 3 which is not a definition of convex hull.

Reply:
The reason is to extend the definition of a convex hull to the non-polygon cases, matching Postgis behaviour. E.g. the convex hull of a point is the same point, and the convex hull of a straight line is the same straight line.